# CS-6570 Lecture 4 - Introduction to Numpy and Pandas

**Dylan Zwick**

*Weber State University*

I decided we shouldn't get too far in our class without a quick review of the two great data science workhorse Python libraries - NumPy (which is short for "Numerical Python") and Pandas (which is technically short for "panel data", but that's more an artifact of history). NumPy is a library for numerical computation, and Pandas is a library for data management.

These are very important libraries, and becoming comfortable with them is a major sign of maturity as a professional data scientist. We will not have the time to dive into them deeply, and this will just be an intro.

We'll start by importing the two libraries. They are almost *always* abbreviated an "np" and "pd", respectively.

In [1]:
import numpy as np # Numpy is almost always abbreviated "np"
import pandas as pd # Pandas is almost always abbreviated "pd"

## Numpy

The basic object of interest in NumPy is the array, or the *ndarray*, which is an efficient, multidimensional array optimized for storage and computation. NumPy also comes with tools for reading/writing array data to disk and some nice built-in math functions, plus tools for integrating with C libraries.

The basic idea behind NumPy is that, while Python is great for doing many things, it's not an inherently optimized language for handling large-scale numeric computations or storing large data files. Instead of requiring that data analysts needing to work with such things do so using a different language, NumPy provides tools that, essentially, translate basic numeric analysis needs into much more efficient, lower-level (C-style) implementations.

To give you an idea of the performance difference, check out the following:

In [2]:
my_arr = np.arange(1000000)
my_list = list(range(1000000))

In [3]:
%timeit my_arr2 = my_arr*2

657 µs ± 54.7 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [4]:
%timeit my_list2 = [x * 2 for x in my_list]

21.6 ms ± 422 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


That's quite the difference! Generally, NumPy-based algorithms are 10 to 100 times faster (or more!) than their pure Python counterparts and use significantly less memory. So, if you're writing for loops to go through lists for large numeric computations - you're probably doing it wrong.

### The NumPy ndarray: A Multidimensional Array Object

The basic data object in NumPy is the *ndarray* ($n$-dimensional array), and applying standard arithmetic operations to these arrays uses syntax very similar to standard arithmetic.

For example, let's create and array:

In [5]:
data = np.array([[1, 4.2, 7], [5, 8, 2.71]])
data

array([[1.  , 4.2 , 7.  ],
       [5.  , 8.  , 2.71]])

We can multiply every element in the array by 2 with the following:

In [6]:
data * 2

array([[ 2.  ,  8.4 , 14.  ],
       [10.  , 16.  ,  5.42]])

Or, we can get the same result by adding the array to itself:

In [7]:
data + data

array([[ 2.  ,  8.4 , 14.  ],
       [10.  , 16.  ,  5.42]])

We could even square every element in the array:

In [8]:
data ** 2

array([[ 1.    , 17.64  , 49.    ],
       [25.    , 64.    ,  7.3441]])

Again, these operations are generally quite fast relative to trying to implement them with a for loop.

#### Data Types for ndarrays

In general, an ndarray is  for homogeneous data; in other words, all the elements must be the same type. Every array also has a *shape*, which is a tuple indicating the size of each dimension, and a *dtype*, which describes the data type of the array.

In [9]:
data.shape

(2, 3)

In [10]:
data.dtype

dtype('float64')

How do we create ndarrays? The easiest way is with the *array* function used above, where we just specify the elements in a list or nested lists. Unless explicitly told otherwise, the array function will try to infer a good data type for the array based upon its inputs.

In addition to the *array* function, there are a number of other functions for creating new arrays. For example:

In [11]:
np.zeros(10)

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [12]:
np.zeros((2,3))

array([[0., 0., 0.],
       [0., 0., 0.]])

In [13]:
np.ones((4,3))

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

The function *arange* is like the *range* function is Python, except it creates a NumPy array instead of a list.

In [14]:
np.arange(12)

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [31]:
np.arange(3,11)

array([ 3,  4,  5,  6,  7,  8,  9, 10])

In [32]:
np.arange(3,11,3)

array([3, 6, 9])

We can, if desired, explicitly convert or *cast* an array from one data type to another using the *astype* method:

In [34]:
arr = np.array([1,2,3,4,5])
arr.dtype

dtype('int64')

In [35]:
float_arr = arr.astype(np.float64)
float_arr

array([1., 2., 3., 4., 5.])

If we go the other way, from floats to ints, that will not generate an error, but the decimal parts will be truncated.

In [37]:
arr = np.array([2.3, 5.0, 4.999, 2.71828])
arr.dtype

dtype('float64')

In [38]:
int_arr = arr.astype(np.int64)
int_arr

array([2, 5, 4, 2])

We can convert strings that make sense as numbers to numbers:

In [40]:
arr = np.array(["2.3", "1", "42"])
arr.dtype

dtype('<U3')

In [41]:
float_arr = arr.astype(np.float64)
float_arr

array([ 2.3,  1. , 42. ])

But, this will cause an error if the strings don't make sense as numbers:

In [16]:
arr = np.array(["2.3", "1", "42", "Shrubbery"])
arr

array(['2.3', '1', '42', 'Shrubbery'], dtype='<U9')

In [17]:
float_arr = arr.astype(np.float64)

ValueError: could not convert string to float: 'Shrubbery'

The datatype for an array can also be specified when you created it:


In [18]:
arr = np.array([1, 2, 3])
arr

array([1, 2, 3])

In [19]:
arr = np.array([1, 2, 3], dtype=np.float64)
arr

array([1., 2., 3.])

#### Arithmetic with NumPy Arrays

Arrays enable you to express batch operations on data without writing any for loops. In NumPy this is called "vectorization". Any arithmetic operations between equal-sized arrays apply the operation element-wise:

In [20]:
arr = np.array([[1,2,3],[3,2,1]])
arr

array([[1, 2, 3],
       [3, 2, 1]])

In [21]:
arr * arr

array([[1, 4, 9],
       [9, 4, 1]])

In [22]:
arr + 2 * arr

array([[3, 6, 9],
       [9, 6, 3]])

#### Basic indexing and slicing

Selecting a subset of an array is a somewhat deep topic that we'll only touch on here. One-dimensional arrays on the surface act similarly to Python lists:

In [23]:
arr = np.arange(3,10)
arr

array([3, 4, 5, 6, 7, 8, 9])

In [24]:
arr[1]

4

In [25]:
arr[2:5]

array([5, 6, 7])

In [26]:
arr[2:5] = 12
arr

array([ 3,  4, 12, 12, 12,  8,  9])

As you can see, assigning a scalar value to a slice propagates (or *broadcasts*) the value to the entire selection.

An important first  distinction here is that array slices are views **on the original array**, which means any modification to them is reflected in the source array.

For example:

In [27]:
arr_slice = arr[2:5]
arr_slice

array([12, 12, 12])

In [28]:
arr_slice[1] = 42

In [29]:
arr

array([ 3,  4, 12, 42, 12,  8,  9])

This might seem surprising. The idea behind this is that NumPy has been designed to work with very large arrays, and you can image that lots of copying of big arrays could lead to performance and memory problems.

If you do want to create a copy and not just a view, you can do so with the *copy* function:

In [157]:
arr = np.arange(10)
arr

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [158]:
arr_slice = arr[3:6]
arr_slice

array([3, 4, 5])

In [159]:
arr_copy = arr[3:6].copy()
arr_copy

array([3, 4, 5])

In [160]:
arr_slice[0] = 23
arr_copy[1] = 42
arr

array([ 0,  1,  2, 23,  4,  5,  6,  7,  8,  9])

For higher dimensional arrays there are more options.

I a two-dimensional array, the elements at each index are no longer scalars but rather one-dimensional arrays:

In [162]:
arr2d = np.array([[1,2,3],[4,5,6],[7,8,9]])
arr2d[2]

array([7, 8, 9])

If we wanted to access the third element of the first array, we could do so either recursively:

In [164]:
arr2d[0][2]

3

Or, using a comma-separated list:

In [166]:
arr2d[0,2]

3

If we assign a scalar value to an entire array, it broadcasts that value to every entry:

In [168]:
arr2d[0]

array([1, 2, 3])

In [169]:
arr2d[0] = 42
arr2d

array([[42, 42, 42],
       [ 4,  5,  6],
       [ 7,  8,  9]])

Multiple slices can be passed to multi-dimensional arrays just like we can pass multiple indices.

In [171]:
arr2d[:2]

array([[42, 42, 42],
       [ 4,  5,  6]])

In [172]:
arr2d[:2,1:]

array([[42, 42],
       [ 5,  6]])

By mixing indexes and slices, you can get lower dimensional slices:

In [174]:
arr2d[:2,2]

array([42,  6])

Keep in mind, these slices are *views*

In [176]:
arr2d[:2,2] = 801
arr2d

array([[ 42,  42, 801],
       [  4,   5, 801],
       [  7,   8,   9]])

We can index the elements in an array with **boolean indexing**. For example, suppose we have the following array of names:

In [178]:
names = np.array(["Bob", "Joe", "Will", "Bob", "Will", "Joe", "Joe"])
data = np.array([[4,7], [0,2], [-5,6], [0,0], [1,2], [-12,-4], [3,4]])

In [179]:
names

array(['Bob', 'Joe', 'Will', 'Bob', 'Will', 'Joe', 'Joe'], dtype='<U4')

In [180]:
data

array([[  4,   7],
       [  0,   2],
       [ -5,   6],
       [  0,   0],
       [  1,   2],
       [-12,  -4],
       [  3,   4]])

Suppose each name corresponds with a row in the *data* array and we wanted to select all the rows with the corresponding name "Bob". Like arithmetic operations, comparisons (like ==) with arrays are also vectorized. So, this command produces a Boolean array:

In [182]:
names == "Bob"

array([ True, False, False,  True, False, False, False])

If we pass this as an index to our array we get:

In [184]:
data[names=="Bob"]

array([[4, 7],
       [0, 0]])

Note the Boolean array must be of the same length as the array axis it's indexing.

You can even mix and match Boolean arrays with slices:

In [186]:
data[names == "Bob", 1:]

array([[7],
       [0]])

In [187]:
data[names == "Bob", 1]

array([7, 0])

We can also use the standard *and* (&), *or* (|), and *not* (~) operations.

In [189]:
data[names != "Bob"]

array([[  0,   2],
       [ -5,   6],
       [  1,   2],
       [-12,  -4],
       [  3,   4]])

In [190]:
data[~(names == "Bob")]

array([[  0,   2],
       [ -5,   6],
       [  1,   2],
       [-12,  -4],
       [  3,   4]])

In [191]:
data[(names == "Bob") | (names == "Will")]

array([[ 4,  7],
       [-5,  6],
       [ 0,  0],
       [ 1,  2]])

Note selecting data from an array by Boolean indexing and assigning the results to a new variable creates a *copy* of the data. I know it's a bit confusing when a copy is created and when it's not. Sorry, I didn't make the rules.

Fancy indexing is a term adopted by NumPy to describe indexing using integer arrays. Sorry, I didn't make the terminology either. For example:

In [194]:
arr = np.array([[1,2,3],[4,5,6],[7,8,9],[10,11,12]])
arr

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [195]:
arr[[1,0,2]]

array([[4, 5, 6],
       [1, 2, 3],
       [7, 8, 9]])

Using negative indices selects rows from the end:

In [197]:
arr[[-1,-2]]

array([[10, 11, 12],
       [ 7,  8,  9]])

Passing multiple index arrays selects an array of elements corresponding to each tuple of indices:

In [199]:
arr[[1,0,2],[2,0,1]]

array([6, 1, 8])

Fancy indexing, unlike slicing, always copies the data into a new array when assigning the results to a new variable. Again, not my rules.

Finally, we can transpose an array with the *T* method:

In [202]:
arr

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [203]:
arr.T

array([[ 1,  4,  7, 10],
       [ 2,  5,  8, 11],
       [ 3,  6,  9, 12]])

## Pandas

For exploring the pandas library, we'll play around with the "royal line" dataset, which was created from public sources and contains family history information about Elizabeth II, the Queen of England at the time the dataset was compiled. Note a few things about the command below:
* It uses the read_csv command from pandas, which is used to read in "comma separated value" files. This is a very common format for storing tabular data, as it's not tied to a particular program like, for example, Excel files are. However, Pandas also has functionality for reading in pretty much any type of data format commonly found in practice.
* I'm accessing the .csv file from my Google drive. This means you all don't need to download it to access it.
* The basic data object in Pandas is a "dataframe", which is created by the read_csv command. Typically, a dataframe is denoted with the abbreviation "df", which we do here.

In [30]:
# This reads in the data from a .csv file on Google Drive
url = 'https://drive.google.com/file/d/1hLUH7s5WPJUfWQrAWQmGeyST4wvTnn7m/view?usp=sharing'
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
df = pd.read_csv(path)

We can then take a look at this dataframe using the "print" command, which will by default print the first five and last five rows of the dataframe.

In [31]:
print(df)

        ID               first_name last_name sex             title  \
0        1                 Victoria   Hanover   F  Queen of England   
1        2  Albert Augustus Charles       NaN   M            Prince   
2        3   Victoria Adelaide Mary       NaN   F    Princess Royal   
3        4               Edward_VII    Wettin   M   King of England   
4        5          Alice Maud Mary       NaN   F          Princess   
...    ...                      ...       ...  ..               ...   
3004  3005                    Emily   Scobell   F               NaN   
3005  3006             John Sanford   Scobell   M               Sir   
3006  3007                    James  Cartland   M               NaN   
3007  3008                    Flora       NaN   F               NaN   
3008  3009                      NaN  Cartland   F               NaN   

       birth_date                          birth_place   death_date  \
0     24 MAY 1819     Kensington,Palace,London,England  22 JAN 1901   
1    

If we just want to check out the first $n$ values, we can use the "head" function:

In [32]:
print(df.head())
print(df.head(12))

   ID               first_name last_name sex             title   birth_date  \
0   1                 Victoria   Hanover   F  Queen of England  24 MAY 1819   
1   2  Albert Augustus Charles       NaN   M            Prince  26 AUG 1819   
2   3   Victoria Adelaide Mary       NaN   F    Princess Royal  21 NOV 1840   
3   4               Edward_VII    Wettin   M   King of England   9 NOV 1841   
4   5          Alice Maud Mary       NaN   F          Princess  25 APR 1843   

                           birth_place   death_date  \
0     Kensington,Palace,London,England  22 JAN 1901   
1  Schloss Rosenau,Near Coburg,Germany  14 DEC 1861   
2     Buckingham,Palace,London,England   5 AUG 1901   
3     Buckingham,Palace,London,England   6 MAY 1910   
4     Buckingham,Palace,London,England  14 DEC 1878   

                           death_place  
0  Osborne House,Isle of Wight,England  
1     Windsor Castle,Berkshire,England  
2   Friedrichshof,Near,Kronberg,Taunus  
3     Buckingham,Palace,London

And similarly the "tail" function returns the last $n$ values:

In [212]:
print(df.tail(15))

        ID             first_name last_name sex  title  birth_date  \
2994  2995       Bertram (Bertie)  Cartland   M  Major         NaN   
2995  2996  Mary Hamilton (Polly)   Scobell   F    NaN  5 SEP 1877   
2996  2997                 Ronald  Cartland   M    NaN  3 JAN 1907   
2997  2998         Anthony (Tony)  Cartland   M    NaN  4 JAN 1912   
2998  2999                  Edith  Palairet   F    NaN         NaN   
2999  3000              Mary Anne  Hamilton   F    NaN         NaN   
3000  3001                 Andrew  Hamilton   M    NaN         NaN   
3001  3002                 George   Scobell   M    NaN         NaN   
3002  3003               Melloney   Scobell   F    NaN         NaN   
3003  3004                    NaN   Scobell   M    NaN         NaN   
3004  3005                  Emily   Scobell   F    NaN         NaN   
3005  3006           John Sanford   Scobell   M    Sir        1879   
3006  3007                  James  Cartland   M    NaN         NaN   
3007  3008          

You may have noticed in the printed dataframe an additional column of numbers located to the far left of the data. For instance, if we just call df.head() with the default option (which is 5), we get:

In [214]:
print(df.head())
df.head()

   ID               first_name last_name sex             title   birth_date  \
0   1                 Victoria   Hanover   F  Queen of England  24 MAY 1819   
1   2  Albert Augustus Charles       NaN   M            Prince  26 AUG 1819   
2   3   Victoria Adelaide Mary       NaN   F    Princess Royal  21 NOV 1840   
3   4               Edward_VII    Wettin   M   King of England   9 NOV 1841   
4   5          Alice Maud Mary       NaN   F          Princess  25 APR 1843   

                           birth_place   death_date  \
0     Kensington,Palace,London,England  22 JAN 1901   
1  Schloss Rosenau,Near Coburg,Germany  14 DEC 1861   
2     Buckingham,Palace,London,England   5 AUG 1901   
3     Buckingham,Palace,London,England   6 MAY 1910   
4     Buckingham,Palace,London,England  14 DEC 1878   

                           death_place  
0  Osborne House,Isle of Wight,England  
1     Windsor Castle,Berkshire,England  
2   Friedrichshof,Near,Kronberg,Taunus  
3     Buckingham,Palace,London

ID               first_name last_name sex             title   birth_date  \
0   1                 Victoria   Hanover   F  Queen of England  24 MAY 1819   
1   2  Albert Augustus Charles       NaN   M            Prince  26 AUG 1819   
2   3   Victoria Adelaide Mary       NaN   F    Princess Royal  21 NOV 1840   
3   4               Edward_VII    Wettin   M   King of England   9 NOV 1841   
4   5          Alice Maud Mary       NaN   F          Princess  25 APR 1843   

                           birth_place   death_date  \
0     Kensington,Palace,London,England  22 JAN 1901   
1  Schloss Rosenau,Near Coburg,Germany  14 DEC 1861   
2     Buckingham,Palace,London,England   5 AUG 1901   
3     Buckingham,Palace,London,England   6 MAY 1910   
4     Buckingham,Palace,London,England  14 DEC 1878   

                           death_place  
0  Osborne House,Isle of Wight,England  
1     Windsor Castle,Berkshire,England  
2   Friedrichshof,Near,Kronberg,Taunus  
3     Buckingham,Palace,London,England  
4                  Darmstadt,,,Germany

That first column is an index column created by Pandas for the dataframe. It starts at $0$ and enumerates from there. Note this column is *not* in our original csv, it's created.

In this example, our original csv already has an index column, called ID, that starts at 1, so this additional data column is a bit redundant. We can specify an index column when we read in the data using the "index_col" parameter in read_csv.

In [216]:
df = pd.read_csv(path, index_col='ID')

Now, the index column is the "ID" column from the original dataset.

In [218]:
df.head()

first_name last_name sex             title   birth_date  \
ID                                                                         
1                  Victoria   Hanover   F  Queen of England  24 MAY 1819   
2   Albert Augustus Charles       NaN   M            Prince  26 AUG 1819   
3    Victoria Adelaide Mary       NaN   F    Princess Royal  21 NOV 1840   
4                Edward_VII    Wettin   M   King of England   9 NOV 1841   
5           Alice Maud Mary       NaN   F          Princess  25 APR 1843   

                            birth_place   death_date  \
ID                                                     
1      Kensington,Palace,London,England  22 JAN 1901   
2   Schloss Rosenau,Near Coburg,Germany  14 DEC 1861   
3      Buckingham,Palace,London,England   5 AUG 1901   
4      Buckingham,Palace,London,England   6 MAY 1910   
5      Buckingham,Palace,London,England  14 DEC 1878   

                            death_place  
ID                                       
1   Osborne House,Isle of Wight,England  
2      Windsor Castle,Berkshire,England  
3    Friedrichshof,Near,Kronberg,Taunus  
4      Buckingham,Palace,London,England  
5                   Darmstadt,,,Germany

If we only wanted to see the "title" column, we could restrict our investigation to that as follows:

In [220]:
df['title'].head()

ID
1    Queen of England
2              Prince
3      Princess Royal
4     King of England
5            Princess
Name: title, dtype: object

We can specify more than one column as follows. Note the *double brackets*. Think of this as the outer brackets accessing the dataframe, while the inner brackets specify a list.

In [222]:
df[['title', 'first_name']].tail()

title    first_name
ID                      
3005   NaN         Emily
3006   Sir  John Sanford
3007   NaN         James
3008   NaN         Flora
3009   NaN           NaN

In [223]:
df.columns

Index(['first_name', 'last_name', 'sex', 'title', 'birth_date', 'birth_place',
       'death_date', 'death_place'],
      dtype='object')

Note this returns an index objects that behaves as an iterable list, so you could, for example, go through the colums with a for loop.

You can also find out more information about a dataframe using the "info" function:

In [225]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3009 entries, 1 to 3009
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   first_name   2984 non-null   object
 1   last_name    1142 non-null   object
 2   sex          2996 non-null   object
 3   title        1398 non-null   object
 4   birth_date   1734 non-null   object
 5   birth_place  486 non-null    object
 6   death_date   1692 non-null   object
 7   death_place  449 non-null    object
dtypes: object(8)
memory usage: 211.6+ KB


### Dropping or Removing Data

There are many reasons why you might want to drop or remove data. For example, it could be that only some data is relevant to your analysis. Or, it could be that some data is insufficient or corrupt, and leaving it in would lead to incorrect conclusions. Also, sometimes certain columns aren't of interest to the analysis in question.

If we want to drop entire columns, we can use the "drop" function:

In [228]:
df.drop(columns = ['birth_place', 'death_place'])

first_name last_name sex             title   birth_date  \
ID                                                                           
1                    Victoria   Hanover   F  Queen of England  24 MAY 1819   
2     Albert Augustus Charles       NaN   M            Prince  26 AUG 1819   
3      Victoria Adelaide Mary       NaN   F    Princess Royal  21 NOV 1840   
4                  Edward_VII    Wettin   M   King of England   9 NOV 1841   
5             Alice Maud Mary       NaN   F          Princess  25 APR 1843   
...                       ...       ...  ..               ...          ...   
3005                    Emily   Scobell   F               NaN          NaN   
3006             John Sanford   Scobell   M               Sir         1879   
3007                    James  Cartland   M               NaN          NaN   
3008                    Flora       NaN   F               NaN          NaN   
3009                      NaN  Cartland   F               NaN          NaN   

       death_date  
ID                 
1     22 JAN 1901  
2     14 DEC 1861  
3      5 AUG 1901  
4      6 MAY 1910  
5     14 DEC 1878  
...           ...  
3005          NaN  
3006          NaN  
3007          NaN  
3008          NaN  
3009  ABT    1911  

[3009 rows x 6 columns]

However, while the dataframe above only has six columns, if we call it again we get:

In [230]:
df

first_name last_name sex             title   birth_date  \
ID                                                                           
1                    Victoria   Hanover   F  Queen of England  24 MAY 1819   
2     Albert Augustus Charles       NaN   M            Prince  26 AUG 1819   
3      Victoria Adelaide Mary       NaN   F    Princess Royal  21 NOV 1840   
4                  Edward_VII    Wettin   M   King of England   9 NOV 1841   
5             Alice Maud Mary       NaN   F          Princess  25 APR 1843   
...                       ...       ...  ..               ...          ...   
3005                    Emily   Scobell   F               NaN          NaN   
3006             John Sanford   Scobell   M               Sir         1879   
3007                    James  Cartland   M               NaN          NaN   
3008                    Flora       NaN   F               NaN          NaN   
3009                      NaN  Cartland   F               NaN          NaN   

                              birth_place   death_date  \
ID                                                       
1        Kensington,Palace,London,England  22 JAN 1901   
2     Schloss Rosenau,Near Coburg,Germany  14 DEC 1861   
3        Buckingham,Palace,London,England   5 AUG 1901   
4        Buckingham,Palace,London,England   6 MAY 1910   
5        Buckingham,Palace,London,England  14 DEC 1878   
...                                   ...          ...   
3005                                  NaN          NaN   
3006                                  NaN          NaN   
3007                                  NaN          NaN   
3008                                  NaN          NaN   
3009                                  NaN  ABT    1911   

                              death_place  
ID                                         
1     Osborne House,Isle of Wight,England  
2        Windsor Castle,Berkshire,England  
3      Friedrichshof,Near,Kronberg,Taunus  
4        Buckingham,Palace,London,England  
5                     Darmstadt,,,Germany  
...                                   ...  
3005                                  NaN  
3006                                  NaN  
3007                                  NaN  
3008                                  NaN  
3009                                  NaN  

[3009 rows x 8 columns]

What?!? I thought we dropped two!

What's going on here is that the drop command creates a new dataframe as its output. It *does not* modify the original dataframe. So, for example, we could say:

In [232]:
df2 = df.drop(columns = ['birth_place', 'death_place'])
df2

first_name last_name sex             title   birth_date  \
ID                                                                           
1                    Victoria   Hanover   F  Queen of England  24 MAY 1819   
2     Albert Augustus Charles       NaN   M            Prince  26 AUG 1819   
3      Victoria Adelaide Mary       NaN   F    Princess Royal  21 NOV 1840   
4                  Edward_VII    Wettin   M   King of England   9 NOV 1841   
5             Alice Maud Mary       NaN   F          Princess  25 APR 1843   
...                       ...       ...  ..               ...          ...   
3005                    Emily   Scobell   F               NaN          NaN   
3006             John Sanford   Scobell   M               Sir         1879   
3007                    James  Cartland   M               NaN          NaN   
3008                    Flora       NaN   F               NaN          NaN   
3009                      NaN  Cartland   F               NaN          NaN   

       death_date  
ID                 
1     22 JAN 1901  
2     14 DEC 1861  
3      5 AUG 1901  
4      6 MAY 1910  
5     14 DEC 1878  
...           ...  
3005          NaN  
3006          NaN  
3007          NaN  
3008          NaN  
3009  ABT    1911  

[3009 rows x 6 columns]

Here, df2 is the dataframe with those two columns dropped, while df is the original, unchanged dataframe.

If we want to actually make the change to the original dataframe, we can do this with the "inplace" parameter.

In [234]:
df.drop(columns = ['birth_place', 'death_place'], inplace = True)
df

first_name last_name sex             title   birth_date  \
ID                                                                           
1                    Victoria   Hanover   F  Queen of England  24 MAY 1819   
2     Albert Augustus Charles       NaN   M            Prince  26 AUG 1819   
3      Victoria Adelaide Mary       NaN   F    Princess Royal  21 NOV 1840   
4                  Edward_VII    Wettin   M   King of England   9 NOV 1841   
5             Alice Maud Mary       NaN   F          Princess  25 APR 1843   
...                       ...       ...  ..               ...          ...   
3005                    Emily   Scobell   F               NaN          NaN   
3006             John Sanford   Scobell   M               Sir         1879   
3007                    James  Cartland   M               NaN          NaN   
3008                    Flora       NaN   F               NaN          NaN   
3009                      NaN  Cartland   F               NaN          NaN   

       death_date  
ID                 
1     22 JAN 1901  
2     14 DEC 1861  
3      5 AUG 1901  
4      6 MAY 1910  
5     14 DEC 1878  
...           ...  
3005          NaN  
3006          NaN  
3007          NaN  
3008          NaN  
3009  ABT    1911  

[3009 rows x 6 columns]

This is the same as:

In [236]:
df = pd.read_csv(path, index_col='ID')
df = df.drop(columns = ['birth_place', 'death_place'])
df

first_name last_name sex             title   birth_date  \
ID                                                                           
1                    Victoria   Hanover   F  Queen of England  24 MAY 1819   
2     Albert Augustus Charles       NaN   M            Prince  26 AUG 1819   
3      Victoria Adelaide Mary       NaN   F    Princess Royal  21 NOV 1840   
4                  Edward_VII    Wettin   M   King of England   9 NOV 1841   
5             Alice Maud Mary       NaN   F          Princess  25 APR 1843   
...                       ...       ...  ..               ...          ...   
3005                    Emily   Scobell   F               NaN          NaN   
3006             John Sanford   Scobell   M               Sir         1879   
3007                    James  Cartland   M               NaN          NaN   
3008                    Flora       NaN   F               NaN          NaN   
3009                      NaN  Cartland   F               NaN          NaN   

       death_date  
ID                 
1     22 JAN 1901  
2     14 DEC 1861  
3      5 AUG 1901  
4      6 MAY 1910  
5     14 DEC 1878  
...           ...  
3005          NaN  
3006          NaN  
3007          NaN  
3008          NaN  
3009  ABT    1911  

[3009 rows x 6 columns]

You can also drop rows by indicating specific indices:

In [238]:
df = pd.read_csv(path, index_col='ID')
df.drop(index=[4,5,6], inplace=True)
df

first_name last_name sex             title   birth_date  \
ID                                                                           
1                    Victoria   Hanover   F  Queen of England  24 MAY 1819   
2     Albert Augustus Charles       NaN   M            Prince  26 AUG 1819   
3      Victoria Adelaide Mary       NaN   F    Princess Royal  21 NOV 1840   
7     Helena Augusta Victoria       NaN   F          Princess  25 MAY 1846   
8     Louise Caroline Alberta       NaN   F          Princess  18 MAR 1848   
...                       ...       ...  ..               ...          ...   
3005                    Emily   Scobell   F               NaN          NaN   
3006             John Sanford   Scobell   M               Sir         1879   
3007                    James  Cartland   M               NaN          NaN   
3008                    Flora       NaN   F               NaN          NaN   
3009                      NaN  Cartland   F               NaN          NaN   

                              birth_place   death_date  \
ID                                                       
1        Kensington,Palace,London,England  22 JAN 1901   
2     Schloss Rosenau,Near Coburg,Germany  14 DEC 1861   
3        Buckingham,Palace,London,England   5 AUG 1901   
7        Buckingham,Palace,London,England   9 JUN 1923   
8        Buckingham,Palace,London,England   3 DEC 1939   
...                                   ...          ...   
3005                                  NaN          NaN   
3006                                  NaN          NaN   
3007                                  NaN          NaN   
3008                                  NaN          NaN   
3009                                  NaN  ABT    1911   

                                   death_place  
ID                                              
1          Osborne House,Isle of Wight,England  
2             Windsor Castle,Berkshire,England  
3           Friedrichshof,Near,Kronberg,Taunus  
7     Schomberg House,Pall Mall,London,England  
8             Kensington,Palace,London,England  
...                                        ...  
3005                                       NaN  
3006                                       NaN  
3007                                       NaN  
3008                                       NaN  
3009                                       NaN  

[3006 rows x 8 columns]

Or, by using df.index, which avoids potential variations in index numbering and always references the first row starting with $0$.

In [240]:
df.drop(df.index[0], inplace=True)
df.drop(df.index[0], inplace=True)
df

first_name last_name sex           title   birth_date  \
ID                                                                         
3      Victoria Adelaide Mary       NaN   F  Princess Royal  21 NOV 1840   
7     Helena Augusta Victoria       NaN   F        Princess  25 MAY 1846   
8     Louise Caroline Alberta       NaN   F        Princess  18 MAR 1848   
9      Arthur William Patrick       NaN   M          Prince   1 MAY 1850   
10      Leopold George Duncan       NaN   M          Prince   7 APR 1853   
...                       ...       ...  ..             ...          ...   
3005                    Emily   Scobell   F             NaN          NaN   
3006             John Sanford   Scobell   M             Sir         1879   
3007                    James  Cartland   M             NaN          NaN   
3008                    Flora       NaN   F             NaN          NaN   
3009                      NaN  Cartland   F             NaN          NaN   

                           birth_place   death_date  \
ID                                                    
3     Buckingham,Palace,London,England   5 AUG 1901   
7     Buckingham,Palace,London,England   9 JUN 1923   
8     Buckingham,Palace,London,England   3 DEC 1939   
9     Buckingham,Palace,London,England  16 JAN 1942   
10    Buckingham,Palace,London,England  28 MAR 1884   
...                                ...          ...   
3005                               NaN          NaN   
3006                               NaN          NaN   
3007                               NaN          NaN   
3008                               NaN          NaN   
3009                               NaN  ABT    1911   

                                   death_place  
ID                                              
3           Friedrichshof,Near,Kronberg,Taunus  
7     Schomberg House,Pall Mall,London,England  
8             Kensington,Palace,London,England  
9                          Bagshot Park,Surrey  
10                                      Cannes  
...                                        ...  
3005                                       NaN  
3006                                       NaN  
3007                                       NaN  
3008                                       NaN  
3009                                       NaN  

[3004 rows x 8 columns]

Each line drops the first row on the dataframe, whatever that first row might be. So, these two lines together drop the first two rows. We can also use standard Python indexing and slicing notation to specify indices here.

The "drop_duplicates" function can be used to drop duplicate rows, while the "dropna" function drops every row that includes at least one NA entry. Be careful with this one, as it could potentially drop a lot of rows!

In [244]:
df.dropna(inplace=True)
df

first_name    last_name sex            title   birth_date  \
ID                                                                             
14                   George_V      Windsor   M  King of England   3 JUN 1865   
31                Edward_VIII      Windsor   M  Duke of Windsor  23 JUN 1894   
32                  George_VI      Windsor   M  King of England  14 DEC 1895   
33                       Mary      Windsor   F   Princess Royal  25 APR 1897   
35    George Edward Alexander      Windsor   M     Duke of Kent  20 DEC 1902   
...                       ...          ...  ..              ...          ...   
1146                   Otto_I  Wittelsbach   M  King of Bavaria  27 APR 1848   
1148  Otto_I Friedrich Ludwig  Wittelsbach   M   King of Greece   1 JUN 1815   
1247                     Mary       Stuart   F   Queen of Scots   7 DEC 1542   
1652                   Louise    Rasmussen   F         Countess  21 APR 1815   
2138                 John_III     Sobieski   M   King of Poland   9 JUN 1624   

                                   birth_place   death_date  \
ID                                                            
14              Marlborough Hse,London,England  20 JAN 1936   
31    White Lodge,Richmond Park,Surrey,England  28 MAY 1972   
32    York Cottage,Sandringham,Norfolk,England   6 FEB 1952   
33    York Cottage,Sandringham,Norfolk,England  28 MAR 1965   
35    York Cottage,Sandringham,Norfolk,England  25 AUG 1942   
...                                        ...          ...   
1146                                    Munich  11 OCT 1916   
1148                          Salzburg,Austria  26 JUL 1867   
1247                       Linlithgow,Scotland   8 FEB 1587   
1652                                Copenhagen   6 MAR 1874   
2138                   Olesko,Now:,Ukraine,SSR  17 JUN 1696   

                            death_place  
ID                                       
14          Sandringham,Norfolk,England  
31                       Paris,,,France  
32          Sandringham,Norfolk,England  
33    Harewood House,Yorkshire,,England  
35                    Morven,,,Scotland  
...                                 ...  
1146                        Near,Munich  
1148                    Bamberg,Germany  
1247                           ,England  
1652                             Cannes  
2138          Wilanow,Nr.,Warsaw,Poland  

[75 rows x 8 columns]

### Adding, Modifying Data, and Mapping

Suppose we have a dataframe with NA values, and instead of dropping them we want to fill them with some values we determine. Here are a few ways to do that:

In [247]:
df = pd.read_csv(path, index_col='ID')
# replace ALL NA entries with a fixed value:
df.fillna(0, inplace=True)
df

first_name last_name sex             title   birth_date  \
ID                                                                           
1                    Victoria   Hanover   F  Queen of England  24 MAY 1819   
2     Albert Augustus Charles         0   M            Prince  26 AUG 1819   
3      Victoria Adelaide Mary         0   F    Princess Royal  21 NOV 1840   
4                  Edward_VII    Wettin   M   King of England   9 NOV 1841   
5             Alice Maud Mary         0   F          Princess  25 APR 1843   
...                       ...       ...  ..               ...          ...   
3005                    Emily   Scobell   F                 0            0   
3006             John Sanford   Scobell   M               Sir         1879   
3007                    James  Cartland   M                 0            0   
3008                    Flora         0   F                 0            0   
3009                        0  Cartland   F                 0            0   

                              birth_place   death_date  \
ID                                                       
1        Kensington,Palace,London,England  22 JAN 1901   
2     Schloss Rosenau,Near Coburg,Germany  14 DEC 1861   
3        Buckingham,Palace,London,England   5 AUG 1901   
4        Buckingham,Palace,London,England   6 MAY 1910   
5        Buckingham,Palace,London,England  14 DEC 1878   
...                                   ...          ...   
3005                                    0            0   
3006                                    0            0   
3007                                    0            0   
3008                                    0            0   
3009                                    0  ABT    1911   

                              death_place  
ID                                         
1     Osborne House,Isle of Wight,England  
2        Windsor Castle,Berkshire,England  
3      Friedrichshof,Near,Kronberg,Taunus  
4        Buckingham,Palace,London,England  
5                     Darmstadt,,,Germany  
...                                   ...  
3005                                    0  
3006                                    0  
3007                                    0  
3008                                    0  
3009                                    0  

[3009 rows x 8 columns]

In [248]:
df = pd.read_csv(path, index_col='ID')
# replace the first 2 NA entries in each column with a fixed value:
df.fillna(0, limit=2, inplace=True)
df

first_name last_name sex             title   birth_date  \
ID                                                                           
1                    Victoria   Hanover   F  Queen of England  24 MAY 1819   
2     Albert Augustus Charles         0   M            Prince  26 AUG 1819   
3      Victoria Adelaide Mary         0   F    Princess Royal  21 NOV 1840   
4                  Edward_VII    Wettin   M   King of England   9 NOV 1841   
5             Alice Maud Mary       NaN   F          Princess  25 APR 1843   
...                       ...       ...  ..               ...          ...   
3005                    Emily   Scobell   F               NaN          NaN   
3006             John Sanford   Scobell   M               Sir         1879   
3007                    James  Cartland   M               NaN          NaN   
3008                    Flora       NaN   F               NaN          NaN   
3009                      NaN  Cartland   F               NaN          NaN   

                              birth_place   death_date  \
ID                                                       
1        Kensington,Palace,London,England  22 JAN 1901   
2     Schloss Rosenau,Near Coburg,Germany  14 DEC 1861   
3        Buckingham,Palace,London,England   5 AUG 1901   
4        Buckingham,Palace,London,England   6 MAY 1910   
5        Buckingham,Palace,London,England  14 DEC 1878   
...                                   ...          ...   
3005                                  NaN          NaN   
3006                                  NaN          NaN   
3007                                  NaN          NaN   
3008                                  NaN          NaN   
3009                                  NaN  ABT    1911   

                              death_place  
ID                                         
1     Osborne House,Isle of Wight,England  
2        Windsor Castle,Berkshire,England  
3      Friedrichshof,Near,Kronberg,Taunus  
4        Buckingham,Palace,London,England  
5                     Darmstadt,,,Germany  
...                                   ...  
3005                                  NaN  
3006                                  NaN  
3007                                  NaN  
3008                                  NaN  
3009                                  NaN  

[3009 rows x 8 columns]

In [249]:
df = pd.read_csv(path, index_col='ID')
# replace ALL NA first names with a fixed value:
df['first_name'].fillna('no first name', inplace=True)  
df

first_name last_name sex             title   birth_date  \
ID                                                                           
1                    Victoria   Hanover   F  Queen of England  24 MAY 1819   
2     Albert Augustus Charles       NaN   M            Prince  26 AUG 1819   
3      Victoria Adelaide Mary       NaN   F    Princess Royal  21 NOV 1840   
4                  Edward_VII    Wettin   M   King of England   9 NOV 1841   
5             Alice Maud Mary       NaN   F          Princess  25 APR 1843   
...                       ...       ...  ..               ...          ...   
3005                    Emily   Scobell   F               NaN          NaN   
3006             John Sanford   Scobell   M               Sir         1879   
3007                    James  Cartland   M               NaN          NaN   
3008                    Flora       NaN   F               NaN          NaN   
3009            no first name  Cartland   F               NaN          NaN   

                              birth_place   death_date  \
ID                                                       
1        Kensington,Palace,London,England  22 JAN 1901   
2     Schloss Rosenau,Near Coburg,Germany  14 DEC 1861   
3        Buckingham,Palace,London,England   5 AUG 1901   
4        Buckingham,Palace,London,England   6 MAY 1910   
5        Buckingham,Palace,London,England  14 DEC 1878   
...                                   ...          ...   
3005                                  NaN          NaN   
3006                                  NaN          NaN   
3007                                  NaN          NaN   
3008                                  NaN          NaN   
3009                                  NaN  ABT    1911   

                              death_place  
ID                                         
1     Osborne House,Isle of Wight,England  
2        Windsor Castle,Berkshire,England  
3      Friedrichshof,Near,Kronberg,Taunus  
4        Buckingham,Palace,London,England  
5                     Darmstadt,,,Germany  
...                                   ...  
3005                                  NaN  
3006                                  NaN  
3007                                  NaN  
3008                                  NaN  
3009                                  NaN  

[3009 rows x 8 columns]

In [250]:
df = pd.read_csv(path, index_col='ID')
# replace specific columns with specific values:
values = {'first_name': 'no_first_name', 'last_name': 'no_last_name', 'sex': 'no_sex', 'title': 'no_title', 'birth_date': 'no_birth_date', 'birth_place': 'no_birth_place', 'death_date': 'no_death_date', 'death_place': 'no_death_place'}
df.fillna(value=values, inplace=True)
df

first_name     last_name sex             title  \
ID                                                                  
1                    Victoria       Hanover   F  Queen of England   
2     Albert Augustus Charles  no_last_name   M            Prince   
3      Victoria Adelaide Mary  no_last_name   F    Princess Royal   
4                  Edward_VII        Wettin   M   King of England   
5             Alice Maud Mary  no_last_name   F          Princess   
...                       ...           ...  ..               ...   
3005                    Emily       Scobell   F          no_title   
3006             John Sanford       Scobell   M               Sir   
3007                    James      Cartland   M          no_title   
3008                    Flora  no_last_name   F          no_title   
3009            no_first_name      Cartland   F          no_title   

         birth_date                          birth_place     death_date  \
ID                                                                        
1       24 MAY 1819     Kensington,Palace,London,England    22 JAN 1901   
2       26 AUG 1819  Schloss Rosenau,Near Coburg,Germany    14 DEC 1861   
3       21 NOV 1840     Buckingham,Palace,London,England     5 AUG 1901   
4        9 NOV 1841     Buckingham,Palace,London,England     6 MAY 1910   
5       25 APR 1843     Buckingham,Palace,London,England    14 DEC 1878   
...             ...                                  ...            ...   
3005  no_birth_date                       no_birth_place  no_death_date   
3006           1879                       no_birth_place  no_death_date   
3007  no_birth_date                       no_birth_place  no_death_date   
3008  no_birth_date                       no_birth_place  no_death_date   
3009  no_birth_date                       no_birth_place    ABT    1911   

                              death_place  
ID                                         
1     Osborne House,Isle of Wight,England  
2        Windsor Castle,Berkshire,England  
3      Friedrichshof,Near,Kronberg,Taunus  
4        Buckingham,Palace,London,England  
5                     Darmstadt,,,Germany  
...                                   ...  
3005                       no_death_place  
3006                       no_death_place  
3007                       no_death_place  
3008                       no_death_place  
3009                       no_death_place  

[3009 rows x 8 columns]

In [251]:
df = pd.read_csv(path, index_col='ID')
# ffill and pad: from first row to last row, propagate the most recent row that # is not an NA forward until next valid row
df.fillna(method='ffill', inplace=True)  
df

first_name last_name sex             title   birth_date  \
ID                                                                           
1                    Victoria   Hanover   F  Queen of England  24 MAY 1819   
2     Albert Augustus Charles   Hanover   M            Prince  26 AUG 1819   
3      Victoria Adelaide Mary   Hanover   F    Princess Royal  21 NOV 1840   
4                  Edward_VII    Wettin   M   King of England   9 NOV 1841   
5             Alice Maud Mary    Wettin   F          Princess  25 APR 1843   
...                       ...       ...  ..               ...          ...   
3005                    Emily   Scobell   F             Major   4 JAN 1912   
3006             John Sanford   Scobell   M               Sir         1879   
3007                    James  Cartland   M               Sir         1879   
3008                    Flora  Cartland   F               Sir         1879   
3009                    Flora  Cartland   F               Sir         1879   

                              birth_place   death_date  \
ID                                                       
1        Kensington,Palace,London,England  22 JAN 1901   
2     Schloss Rosenau,Near Coburg,Germany  14 DEC 1861   
3        Buckingham,Palace,London,England   5 AUG 1901   
4        Buckingham,Palace,London,England   6 MAY 1910   
5        Buckingham,Palace,London,England  14 DEC 1878   
...                                   ...          ...   
3005                       Florence,Italy  BEF    1877   
3006                       Florence,Italy  BEF    1877   
3007                       Florence,Italy  BEF    1877   
3008                       Florence,Italy  BEF    1877   
3009                       Florence,Italy  ABT    1911   

                              death_place  
ID                                         
1     Osborne House,Isle of Wight,England  
2        Windsor Castle,Berkshire,England  
3      Friedrichshof,Near,Kronberg,Taunus  
4        Buckingham,Palace,London,England  
5                     Darmstadt,,,Germany  
...                                   ...  
3005                     Nr Cassel,France  
3006                     Nr Cassel,France  
3007                     Nr Cassel,France  
3008                     Nr Cassel,France  
3009                     Nr Cassel,France  

[3009 rows x 8 columns]

In [252]:
df = pd.read_csv(path, index_col='ID')
# bfill and backfill: like ffill, except from last row to first row
df.fillna(method='bfill', inplace=True)
df

first_name last_name sex             title   birth_date  \
ID                                                                           
1                    Victoria   Hanover   F  Queen of England  24 MAY 1819   
2     Albert Augustus Charles    Wettin   M            Prince  26 AUG 1819   
3      Victoria Adelaide Mary    Wettin   F    Princess Royal  21 NOV 1840   
4                  Edward_VII    Wettin   M   King of England   9 NOV 1841   
5             Alice Maud Mary   Windsor   F          Princess  25 APR 1843   
...                       ...       ...  ..               ...          ...   
3005                    Emily   Scobell   F               Sir         1879   
3006             John Sanford   Scobell   M               Sir         1879   
3007                    James  Cartland   M               NaN          NaN   
3008                    Flora  Cartland   F               NaN          NaN   
3009                      NaN  Cartland   F               NaN          NaN   

                              birth_place   death_date  \
ID                                                       
1        Kensington,Palace,London,England  22 JAN 1901   
2     Schloss Rosenau,Near Coburg,Germany  14 DEC 1861   
3        Buckingham,Palace,London,England   5 AUG 1901   
4        Buckingham,Palace,London,England   6 MAY 1910   
5        Buckingham,Palace,London,England  14 DEC 1878   
...                                   ...          ...   
3005                                  NaN  ABT    1911   
3006                                  NaN  ABT    1911   
3007                                  NaN  ABT    1911   
3008                                  NaN  ABT    1911   
3009                                  NaN  ABT    1911   

                              death_place  
ID                                         
1     Osborne House,Isle of Wight,England  
2        Windsor Castle,Berkshire,England  
3      Friedrichshof,Near,Kronberg,Taunus  
4        Buckingham,Palace,London,England  
5                     Darmstadt,,,Germany  
...                                   ...  
3005                                  NaN  
3006                                  NaN  
3007                                  NaN  
3008                                  NaN  
3009                                  NaN  

[3009 rows x 8 columns]

We can also create new columns from existing ones. For example:

In [254]:
df = pd.read_csv(path, index_col='ID')
df['full_name'] = df['first_name'] + ' ' + df['last_name']
df

first_name last_name sex             title   birth_date  \
ID                                                                           
1                    Victoria   Hanover   F  Queen of England  24 MAY 1819   
2     Albert Augustus Charles       NaN   M            Prince  26 AUG 1819   
3      Victoria Adelaide Mary       NaN   F    Princess Royal  21 NOV 1840   
4                  Edward_VII    Wettin   M   King of England   9 NOV 1841   
5             Alice Maud Mary       NaN   F          Princess  25 APR 1843   
...                       ...       ...  ..               ...          ...   
3005                    Emily   Scobell   F               NaN          NaN   
3006             John Sanford   Scobell   M               Sir         1879   
3007                    James  Cartland   M               NaN          NaN   
3008                    Flora       NaN   F               NaN          NaN   
3009                      NaN  Cartland   F               NaN          NaN   

                              birth_place   death_date  \
ID                                                       
1        Kensington,Palace,London,England  22 JAN 1901   
2     Schloss Rosenau,Near Coburg,Germany  14 DEC 1861   
3        Buckingham,Palace,London,England   5 AUG 1901   
4        Buckingham,Palace,London,England   6 MAY 1910   
5        Buckingham,Palace,London,England  14 DEC 1878   
...                                   ...          ...   
3005                                  NaN          NaN   
3006                                  NaN          NaN   
3007                                  NaN          NaN   
3008                                  NaN          NaN   
3009                                  NaN  ABT    1911   

                              death_place             full_name  
ID                                                               
1     Osborne House,Isle of Wight,England      Victoria Hanover  
2        Windsor Castle,Berkshire,England                   NaN  
3      Friedrichshof,Near,Kronberg,Taunus                   NaN  
4        Buckingham,Palace,London,England     Edward_VII Wettin  
5                     Darmstadt,,,Germany                   NaN  
...                                   ...                   ...  
3005                                  NaN         Emily Scobell  
3006                                  NaN  John Sanford Scobell  
3007                                  NaN        James Cartland  
3008                                  NaN                   NaN  
3009                                  NaN                   NaN  

[3009 rows x 9 columns]

This illustrates a problem. Anytime we have an NaN value, the string concatenation is also NaN. How could we get around this? Well, we could create our own specific function that handles this, and then apply that to our dataframe:

In [256]:
def create_full_name(row):
    if isinstance(row['first_name'], str) and isinstance(row['last_name'], str):  # both first_name and last_name are strings
        result = row['first_name'] + ' ' + row['last_name']
    elif isinstance(row['first_name'], str):  # only first_name is a string
        result = row['first_name']
    elif isinstance(row['last_name'], str):  # only last_name is a string
        result = row['last_name']
    else:  # neither first_name nor last_name are strings, they are both NaN
        result = np.nan 
    return result

df = pd.read_csv(path, index_col='ID')

df['full_name'] = df.apply(create_full_name, axis=1)
df

first_name last_name sex             title   birth_date  \
ID                                                                           
1                    Victoria   Hanover   F  Queen of England  24 MAY 1819   
2     Albert Augustus Charles       NaN   M            Prince  26 AUG 1819   
3      Victoria Adelaide Mary       NaN   F    Princess Royal  21 NOV 1840   
4                  Edward_VII    Wettin   M   King of England   9 NOV 1841   
5             Alice Maud Mary       NaN   F          Princess  25 APR 1843   
...                       ...       ...  ..               ...          ...   
3005                    Emily   Scobell   F               NaN          NaN   
3006             John Sanford   Scobell   M               Sir         1879   
3007                    James  Cartland   M               NaN          NaN   
3008                    Flora       NaN   F               NaN          NaN   
3009                      NaN  Cartland   F               NaN          NaN   

                              birth_place   death_date  \
ID                                                       
1        Kensington,Palace,London,England  22 JAN 1901   
2     Schloss Rosenau,Near Coburg,Germany  14 DEC 1861   
3        Buckingham,Palace,London,England   5 AUG 1901   
4        Buckingham,Palace,London,England   6 MAY 1910   
5        Buckingham,Palace,London,England  14 DEC 1878   
...                                   ...          ...   
3005                                  NaN          NaN   
3006                                  NaN          NaN   
3007                                  NaN          NaN   
3008                                  NaN          NaN   
3009                                  NaN  ABT    1911   

                              death_place                full_name  
ID                                                                  
1     Osborne House,Isle of Wight,England         Victoria Hanover  
2        Windsor Castle,Berkshire,England  Albert Augustus Charles  
3      Friedrichshof,Near,Kronberg,Taunus   Victoria Adelaide Mary  
4        Buckingham,Palace,London,England        Edward_VII Wettin  
5                     Darmstadt,,,Germany          Alice Maud Mary  
...                                   ...                      ...  
3005                                  NaN            Emily Scobell  
3006                                  NaN     John Sanford Scobell  
3007                                  NaN           James Cartland  
3008                                  NaN                    Flora  
3009                                  NaN                 Cartland  

[3009 rows x 9 columns]

This "apply" operation applies the specified function. You could also use Python lambda functions to create a function inline if needed. Note the option "axis = 1" means to process the data row by row. The option "axis = 0" would process the data column by column. For example:

In [258]:
# Create a dataframe that is a 6 x 2 array formed from a list of 12 numbers ordered from 0 to 11.
df = pd.DataFrame(np.arange(12).reshape(6,2), columns = ['column 1', 'column 2'])
print(df)

   column 1  column 2
0         0         1
1         2         3
2         4         5
3         6         7
4         8         9
5        10        11


In [259]:
# Create a new dataframe that takes the maximum value of each column in the dataframe we just created.
new_df = df.apply(lambda x: x.max())
print(new_df)

column 1    10
column 2    11
dtype: int64


There are three main functions used to create or change data in dataframes: apply, map, and applymap.

In [261]:
df = pd.DataFrame(np.arange(8).reshape(4,2), columns = ['column 1', 'column 2'])
print(df)

   column 1  column 2
0         0         1
1         2         3
2         4         5
3         6         7


In [262]:
print(df.apply(np.max))

column 1    6
column 2    7
dtype: int64


In [263]:
print(df.apply(np.max, axis = 1))

0    1
1    3
2    5
3    7
dtype: int64


In [264]:
print(df['column 1'].map(lambda x: x*2))

0     0
1     4
2     8
3    12
Name: column 1, dtype: int64


In [265]:
df['column 1'].map(lambda x: x*2)
df

column 1  column 2
0         0         1
1         2         3
2         4         5
3         6         7

In [266]:
print(df.applymap(lambda x: x*2))

   column 1  column 2
0         0         2
1         4         6
2         8        10
3        12        14


### Changing Datatypes of Series or Columns

The datatypes for our "royal_line" examples have all been 'objects' because every column has had data that's been interpreted as a string. This is a general, default datatype that is quite encompassing in what it can handle. However, there are some functions, like maximum or average, that make sense for certain types of numeric data, but not for general data, and if we try to apply these functions to objects we'll have a bad time.

In [269]:
# Let's create a simple dataframe with three columns containing different types of data:
df = pd.DataFrame({'ints': [1,2,3,4], 'strings': ['a','b','c','d'], 'floats': [1.1, '2.2', '3.3', 4]})
print(df)
print(df.dtypes)

   ints strings floats
0     1       a    1.1
1     2       b    2.2
2     3       c    3.3
3     4       d      4
ints        int64
strings    object
floats     object
dtype: object


Here, the second and third column are interpreted as objects because both contained strings (the values 2.2 and 3.3 in the floats column were entered as strings).

To convert these to a different datatype, we can convert a single column, or multiple columns using a dictionary.

In [271]:
df['floats'] = df['floats'].astype(float)
print(df.dtypes)

ints         int64
strings     object
floats     float64
dtype: object


In [272]:
convert_dict = {'ints': int, 'strings': str, 'floats': float}
df = df.astype(convert_dict)
print(df.dtypes)

ints         int64
strings     object
floats     float64
dtype: object


In [273]:
# The following command would also work:
df['ints'] = df['ints'].astype(float)
print(df)
print(df.dtypes)

   ints strings  floats
0   1.0       a     1.1
1   2.0       b     2.2
2   3.0       c     3.3
3   4.0       d     4.0
ints       float64
strings     object
floats     float64
dtype: object


In [274]:
# But this one won't:
df['strings'] = df['strings'].astype(int)

ValueError: invalid literal for int() with base 10: 'a'

Pandas has many builtin conversion functions (to_datetime, to_timedelta, to_numeric, etc...) but you'll sometimes encounter data that's formatted in such a way that it's not possible to immediately convert it to the format you want using one of the builtin functions. To deal with this, sometimes you need to write your own conversion function.

For example, if we check out the 'birth_date' column in our royal_line dataset, we see:

In [312]:
df = pd.read_csv(path, index_col='ID')
print(df['birth_date'])

ID
1       24 MAY 1819
2       26 AUG 1819
3       21 NOV 1840
4        9 NOV 1841
5       25 APR 1843
           ...     
3005            NaN
3006           1879
3007            NaN
3008            NaN
3009            NaN
Name: birth_date, Length: 3009, dtype: object


A lot of NaN. OK, let's remove these and see what we get:

In [314]:
df.dropna(subset = ['birth_date'], inplace=True)
print(df['birth_date'])

ID
1       24 MAY 1819
2       26 AUG 1819
3       21 NOV 1840
4        9 NOV 1841
5       25 APR 1843
           ...     
2989    11 OCT 1937
2996     5 SEP 1877
2997     3 JAN 1907
2998     4 JAN 1912
3006           1879
Name: birth_date, Length: 1734, dtype: object


If we then try to convert these values to datetimes we get:

In [316]:
df['birth_date'] = pd.to_datetime(df['birth_date']) # This fails

OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 1292-01-01 00:00:00 present at position 123

This generates errors due to several issues. First, there are entries in the dataset formatted like the following: ABT 751. This notation means that the family history experts believe the person was born about (ABT) 751. the second is related to an out of bounds nanosecond timestamp error relating to Pandas only supporting approximately 580 years in the range from around 1677 to 2262.

To get around these issue, we'll write and then apply our own function. Note we're not dropping the NaN values here.

In [356]:
def get_year(x):
    if pd.isna(x):
        year_result = np.nan  # if the birth_year is nan then return nan
    else:  # checking a number of edge cases in the data and stripping it out:
        if "ABT" in x:  # for example: ABT  1775  
            x = x[3:]
            x = x.strip()
        if "/" in x:  #  For example: 1775/1776
            x = x[:x.find('/')]
        num_spaces = x.count(' ')
        if num_spaces == 0:  # only has the year
            year_result = int(x)
        elif num_spaces == 1:  # example: FEB 1337
            x = x[x.rfind(' ') + 1:]  # 'rfind' finds the last space. The 'r' stands for 'reverse.'
            if x.isnumeric():
                year_result = int(x)
            else:  # This could happen if there is only a day and month, like '10 JAN'
                year_result = np.nan
        elif num_spaces == 2:  # example: 16 FEB 1337
            x = x[x.rfind(' ') + 1:]  # 'rfind' finds the last space. The 'r' stands for 'reverse.'
            year_result = int(x)
        else:
            year_result = np.nan  # There are a few other strange dates that aren't worth our time to fix, so just return nan for those.
    return year_result

df['birth_year'] = df['birth_date'].map(get_year)

print(df['birth_year'])

ID
1       1819.0
2       1819.0
3       1840.0
4       1841.0
5       1843.0
         ...  
2989    1937.0
2996    1877.0
2997    1907.0
2998    1912.0
3006    1879.0
Name: birth_year, Length: 1734, dtype: float64


### Conditionals in Dataframes and Series

Conditionals are a very useful feature of Pandas which typically produce a Numpy array of Booleans or a Pandas Boolean series.

For example, consider the following code that produces True if the birth_year column (calculated above) is greater than or equal to 1990, and False otherwise.

In [359]:
boolean_mask = (df.birth_year >= 1990)
print(boolean_mask)

ID
1       False
2       False
3       False
4       False
5       False
        ...  
2989    False
2996    False
2997    False
2998    False
3006    False
Name: birth_year, Length: 1734, dtype: bool


We can then use this to, for example, only print the entries for which the boolean is True.

In [361]:
print(df[['first_name', 'last_name', 'birth_year']][boolean_mask])

                   first_name last_name  birth_year
ID                                                 
2958  Eugenie Victoria Helena   Windsor      1990.0
2961                      NaN    Mowatt      1990.0
2963                    Kitty       NaN      1991.0


We can combine Boolean expressions using the logical operators & ("and"), | ("or"), and ~ ("not"). For example:

In [363]:
print(df[(df.birth_year >= 1500) & (df.title.str.contains('Queen'))][['first_name', 'title', 'birth_year']])

                          first_name             title  birth_year
ID                                                                
1                           Victoria  Queen of England      1819.0
27            Victoria Eugenie Ena""    Queen of Spain      1887.0
30                Mary_of_Teck (May)             Queen      1867.0
52       Elizabeth_II Alexandra Mary  Queen of England      1926.0
76                 Sophie of_Prussia   Queen of Greece      1870.0
96    Marie of_Saxe-Coburg and_Gotha  Queen of Romania      1875.0
610                     Mergrethe_II  Queen of Denmark      1940.0
656              Emma of_Netherlands      Queen Regent      1858.0
657        Wilhelmina of_Netherlands             Queen      1880.0
659           Juliana of_Netherlands             Queen      1909.0
661           Beatrix of_Netherlands             Queen      1938.0
682        Maria Cristina of_Austria    Queen of Spain      1858.0
693                             Anne  Queen of England      16

### loc and iloc Functions

One of the most common tasks for data scientists is filtering the information to more efficiently derive actionable insights about the data. We've seen the "head" and "tail" functions, which provide a quick, truncated view of the beginning or end, respectively, of the dataframe or series. But what if you're interested in examining results that are not necessarily at the very beginning or end of the dataset.

For this purpose, the loc function is designed to access rows and columns by label. In contrast, the iloc function is used to access rows and columns by integer value - the "i" stands for "integer".

A quick example of the difference is illustrated below, where both commands do the same thing:

In [366]:
print(df.loc[1])

first_name                                Victoria
last_name                                  Hanover
sex                                              F
title                             Queen of England
birth_date                             24 MAY 1819
birth_place       Kensington,Palace,London,England
death_date                             22 JAN 1901
death_place    Osborne House,Isle of Wight,England
birth_year                                  1819.0
Name: 1, dtype: object


In [367]:
print(df.iloc[0])

first_name                                Victoria
last_name                                  Hanover
sex                                              F
title                             Queen of England
birth_date                             24 MAY 1819
birth_place       Kensington,Palace,London,England
death_date                             22 JAN 1901
death_place    Osborne House,Isle of Wight,England
birth_year                                  1819.0
Name: 1, dtype: object


However, the following will produce an error:

In [369]:
print(df.loc[0])

KeyError: 0

Because there is no row with label 0.

Now, row indices (labels) don't need to be unique. For example:

In [402]:
df = pd.DataFrame(np.arange(10).reshape(5, 2), columns=['A', 'B'], index=['cat', 42, 'stone', 42, 12345])# Five rows each with an associated index
print(df)

       A  B
cat    0  1
42     2  3
stone  4  5
42     6  7
12345  8  9


The index "42" appears twice, and some indices are numbers, while some are strings. Let's look at some examples:

In [404]:
print(df.loc[12345])

A    8
B    9
Name: 12345, dtype: int64


In [405]:
print(df.loc['stone'])

A    4
B    5
Name: stone, dtype: int64


In [406]:
print(df.loc[42])

    A  B
42  2  3
42  6  7


In [439]:
print(df.loc['cat':'stone'])

       A  B
cat    0  1
42     2  3
stone  4  5


In [440]:
print(df.loc[['cat','stone']])

       A  B
cat    0  1
stone  4  5


In [441]:
print(df.loc['stone', 'B'])

5


In [442]:
print(df.loc[df['A'] > 3])

       A  B
stone  4  5
42     6  7
12345  8  9


Now let's take a look at some iloc examples:

In [444]:
print(df.iloc[0])

A    0
B    1
Name: cat, dtype: int64


In [445]:
print(df.iloc[0:3])

       A  B
cat    0  1
42     2  3
stone  4  5


In [446]:
print(df.iloc[[0,2,4]])

       A  B
cat    0  1
stone  4  5
12345  8  9


In [447]:
print(df.iloc[0,1])

1


In [448]:
print(df.iloc[0:3,1])

cat      1
42       3
stone    5
Name: B, dtype: int64


Returning to the royal family history data as an example, let's create a new column named "era". The "era" column signifies if a person was born in one of three distinct time periods: 'ancient', 'middle_years', or 'modern'. The following creates a new column and initially assigns the value 'unknown' to every entry within it:

In [450]:
df = pd.read_csv(path, index_col='ID')
df['birth_year'] = df['birth_date'].map(get_year)
df['era'] = 'unknown'

In [451]:
print(df)

                   first_name last_name sex             title   birth_date  \
ID                                                                           
1                    Victoria   Hanover   F  Queen of England  24 MAY 1819   
2     Albert Augustus Charles       NaN   M            Prince  26 AUG 1819   
3      Victoria Adelaide Mary       NaN   F    Princess Royal  21 NOV 1840   
4                  Edward_VII    Wettin   M   King of England   9 NOV 1841   
5             Alice Maud Mary       NaN   F          Princess  25 APR 1843   
...                       ...       ...  ..               ...          ...   
3005                    Emily   Scobell   F               NaN          NaN   
3006             John Sanford   Scobell   M               Sir         1879   
3007                    James  Cartland   M               NaN          NaN   
3008                    Flora       NaN   F               NaN          NaN   
3009                      NaN  Cartland   F               NaN   

The next question is how to divide the birth years. If we check out their maximum and minimum values, we get:

In [453]:
print(f"The earlier year = {df['birth_year'].min()} and the latest year = {df['birth_year'].max()}.")

The earlier year = 686.0 and the latest year = 1991.0.


So, 686 is the earliest year, and 1991 is the latest. This is a difference of 1991 - 656 = 1305 years, which if we divide by 3 this gives us 435 years per era. So, the "ancient" royals are those born between 686 and 1121, the "middle_years" royals are those born between 1121 and 1555, and the "modern" royals are those born after 1555. (Not all that modern!) We can assign these three eras with the following code:

In [455]:
df.loc[df['birth_year'] < 1122, 'era'] = 'ancient'  # 686 – 1121
df.loc[(df['birth_year'] >= 1122) & (df['birth_year'] <= 1555), 'era'] = 'middle_years'  # 1122 – 1555
df.loc[df['birth_year'] > 1555, 'era'] = 'modern'  # after 1555
print(df)

                   first_name last_name sex             title   birth_date  \
ID                                                                           
1                    Victoria   Hanover   F  Queen of England  24 MAY 1819   
2     Albert Augustus Charles       NaN   M            Prince  26 AUG 1819   
3      Victoria Adelaide Mary       NaN   F    Princess Royal  21 NOV 1840   
4                  Edward_VII    Wettin   M   King of England   9 NOV 1841   
5             Alice Maud Mary       NaN   F          Princess  25 APR 1843   
...                       ...       ...  ..               ...          ...   
3005                    Emily   Scobell   F               NaN          NaN   
3006             John Sanford   Scobell   M               Sir         1879   
3007                    James  Cartland   M               NaN          NaN   
3008                    Flora       NaN   F               NaN          NaN   
3009                      NaN  Cartland   F               NaN   

We could have also done this using a custom function and the "map" utility.

### Reshaping with Pivot, Pivot_Table, Groupby, and Transpose

Frequently it is convenient or informative to restructure data contained in a dataframe, effectively organizing the data into a different shape or format. This section will cover the most common reshaping functions provided by Pandas.

The pivot function addresses the situation in which separate categories of a dataset feature are enumerated and highlighted using a cross tabular format. For example:

In [459]:
df = pd.DataFrame({'Car': [1, 1, 2, 2],
                   'Type': ['new', 'used', 'new', 'used'],
                   'Price': [10, 5, 12, 7]})
print(df)

   Car  Type  Price
0    1   new     10
1    1  used      5
2    2   new     12
3    2  used      7


Calling the pivot function on this dataframe will reframe the data into a more compact and usable format. In the example above, we want to reshape the data such that each car brand is represented on a single row. A use use case for this particular reorganization would be a care salesperson who needs to quickly view all the prices of a given car brand for the different 'Type' categories.

In [461]:
p = df.pivot(index='Car', columns='Type', values='Price')
print(p)

Type  new  used
Car            
1      10     5
2      12     7


The pivot function only works if there is either zero or one entries per cell in the result. Suppose we have the following dataframe:

In [463]:
df = pd.DataFrame({'Car': [1, 1, 2, 2, 2],
                   'Type': ['new', 'used', 'new', 'used', 'used'],
                   'Price': [10, 5, 12, 7, 6]})
print(df)

   Car  Type  Price
0    1   new     10
1    1  used      5
2    2   new     12
3    2  used      7
4    2  used      6


Invoking the pivot function on this dataframe will generate an error:

In [465]:
p = df.pivot(index='Car', columns='Type', values='Price')

ValueError: Index contains duplicate entries, cannot reshape

The reason for this is there are two price entries for the used version of car 2. In this case what we could do is use the "pivot_table" function with an aggregator, which specifies how to combine values when more than 1 occurs.

In [476]:
p = df.pivot_table(index='Car', columns='Type', values='Price', aggfunc='mean')
print(p)

Type   new  used
Car             
1     10.0   5.0
2     12.0   6.5


Pivot tables can result in immensely complex tabular formats with multiple indexes, multiple columns, and various aggregation functions specified. Today, we demonstrate only the basic single-index, single-column case.

Here's another example of a pivot_table using our royal family history dataset:

In [479]:
df = pd.read_csv(path, index_col='ID')
df['birth_year'] = df['birth_date'].map(get_year)

df.dropna(inplace=True, subset=['title', 'sex', 'birth_year'])
p = df.pivot_table(index='title', columns='sex', values='birth_year', aggfunc='mean')
print(p)

sex                    F       M
title                           
Admiral              NaN  1759.0
Admiral Sir          NaN  1881.0
Archduchess       1825.0     NaN
Archduke             NaN  1844.6
Baron                NaN  1901.0
...                  ...     ...
Tsarina           1859.5     NaN
Vicount Althorp      NaN  1964.0
Vicount Linley       NaN  1961.0
Viscount             NaN  1913.8
Viscount Hampden     NaN  1877.5

[171 rows x 2 columns]


Here are two more examples. The first fills blank entries in the resulting pivot table after aggregation with $0$ instead of NaN, and uses two aggregate functions, mean and count.

In [481]:
p = df.pivot_table(index='title', columns='sex', values='birth_year', aggfunc=['mean', 'count'], fill_value=0)
print(p)

                    mean         count   
sex                    F       M     F  M
title                                    
Admiral              0.0  1759.0     0  1
Admiral Sir          0.0  1881.0     0  1
Archduchess       1825.0     0.0     1  0
Archduke             0.0  1844.6     0  5
Baron                0.0  1901.0     0  4
...                  ...     ...   ... ..
Tsarina           1859.5     0.0     2  0
Vicount Althorp      0.0  1964.0     0  1
Vicount Linley       0.0  1961.0     0  1
Viscount             0.0  1913.8     0  5
Viscount Hampden     0.0  1877.5     0  2

[171 rows x 4 columns]


The second is similar to the first, but instead declares two indexes and two columns producing a much more complicated, nested output result.

In [483]:
p = df.pivot_table(index=['title', 'first_name'], columns=['sex', 'last_name'], values='birth_year', aggfunc=['mean', 'count'], fill_value=0)
print(p)

                                                    mean                \
sex                                                    F                 
last_name                                Armstrong-Jones Ashley Baring   
title            first_name                                              
Admiral          Hugh                                  0      0      0   
Admiral Sir      Alexander                             0      0      0   
Baron            Carl Marten                           0      0      0   
                 Nicholas                              0      0      0   
                 Paul                                  0      0      0   
...                                                  ...    ...    ...   
Viscount         George Earl_of_Harewood               0      0      0   
                 Henry George Charles                  0      0      0   
                 Louis de_la_Torre                     0      0      0   
                 Mervyn Powerscourt   

The groupby function's recasting of information is very similar to that of the pivot_table function. In general, the main difference is how the resulting output is shaped. Note it's a common mistake to create a group object without specifying an aggregating function like mean, sum, or std.

Consider the following:

In [485]:
df = pd.DataFrame({'Car': [1, 1, 2, 2, 2],
                   'Type': ['new', 'used', 'new', 'used', 'used'],
                   'Price': [10, 5, 12, 7, 6]})

df

Car  Type  Price
0    1   new     10
1    1  used      5
2    2   new     12
3    2  used      7
4    2  used      6

In [486]:
g = df.groupby(by='Car')
print(g)

That's not particularly helpful. However, if we group by 'Car' and invoke the 'mean' function, we obtain a more useful result.

In [488]:
g = df.groupby(by='Car').mean()
print(g)

        Price
Car          
1    7.500000
2    8.333333


/tmp/ipykernel_546/4121060130.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  g = df.groupby(by='Car').mean()


That warning basically tells us that we tried to take the mean of "type", which doesn't make sense, so it was tossed out. If instead of mean we wanted to use the more robust nunique (a count of the distinct values) we'd get:

In [490]:
g = df.groupby(by='Car').nunique()
print(g)

     Type  Price
Car             
1       2      2
2       2      3


If we wanted to group by both 'Car' and 'Type', and use two different aggregation function, we could do that:

In [492]:
g = df.groupby(by=['Car','Type']).agg(['mean','nunique'])
print(g)

         Price        
          mean nunique
Car Type              
1   new   10.0       1
    used   5.0       1
2   new   12.0       1
    used   6.5       2


The transpose function (or simply T) transposes a dataframe. For example: